src/syn/io/der.rs

```diff
@@ -145,62 +154,245 @@ impl Reader for DerReader {
+        }
+
+        Ok((index, C::read_content(index, self)?)).and_then(|(index, content)| {
+            content.ok_or_else(|| Error::InvalidChoiceIndex(index, C::VARIANT_COUNT))
+        })
     }
 
     #[inline]
     fn read_opt<T: ReadableType>(
         &mut self,
     ) -> Result<Option<<T as ReadableType>::Type>, Self::Error> {
-        unimplemented!()
+        let read_position = self.buffer.read_position;
+        match T::read_value(self) {
+            Ok(result) => Ok(Some(result)),
+            Err(Error::InvalidType) => {
+                self.buffer.read_position = read_position;
+                Ok(None)
+            },
+            Err(err) => Err(err)
+        }
     }
 
     #[inline]
     fn read_number<T: numbers::Number, C: numbers::Constraint<T>>(
         &mut self,
     ) -> Result<T, Self::Error> {
-        unimplemented!()
+        let (class, pc, tag) = self.buffer.read_identifier()?;
+        let length = self.buffer.read_length()?;
+
+        if let Class::Universal = &class {
+            if let (PC::Primitive, 2) = (&pc, &tag) {} else {
+                return Err(Error::InvalidType)
+            }
+        }
+
+        eprintln!("[number] Class = {:#?}, PC = {:#?}, Tag = {:#?}, Length = {:#?}", class, pc, tag, length);
+
+        if let Length::Definite(l) = length {
+            self.buffer.read_i64_number(l).map(T::from_i64)
+        } else {
+            Err(Error::UnsupportedOperation("Indefinite range is not supported in DER".to_string()))
+        }
     }
 
     #[inline]
     fn read_utf8string<C: utf8string::Constraint>(&mut self) -> Result<String, Self::Error> {
-        unimplemented!()
+        let (class, pc, tag) = self.buffer.read_identifier()?;
+        let length = self.buffer.read_length()?;
+
+        eprintln!("[utf8string] Class = {:#?}, PC = {:#?}, Tag = {:#?}, Length = {:#?}", class, pc, tag, length);
+
+        if let Class::Universal = &class {
+            if let 12 = &tag {
+            } else {
+                return Err(Error::InvalidType)
+            }
+        }
+
+        if let Length::Definite(l) = length {
+            let octets = self.buffer.read_octet_string(l)?;
+            String::from_utf8(octets).map_err(|_| Self::Error::InvalidUtf8String)
+        } else {
+            Err(Error::UnsupportedOperation("Indefinite range is not supported in DER".to_string()))
+        }
     }
 
     #[inline]
     fn read_ia5string<C: ia5string::Constraint>(&mut self) -> Result<String, Self::Error> {
-        unimplemented!()
+        let (class, pc, tag) = self.buffer.read_identifier()?;
+        let length = self.buffer.read_length()?;
+
+        eprintln!("[ia5string] Class = {:#?}, PC = {:#?}, Tag = {:#?}, Length = {:#?}", class, pc, tag, length);
+
+        if let Class::Universal = &class {
+            if let 22 = &tag {
+            } else {
+                return Err(Error::InvalidType)
+            }
+        }
```

Comment:
With `asn1rs::model::Type` this can be collapsed to
```rust
if matches!(tag, Tag::Universal(n) if n != 22) {
    return Err(Error::InvalidType);
}
```
or more preferably
```rust
if Tag::DEFAULT_IA5_STRING != tag {
    return Err(Error::InvalidType);
}
```